src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1173 @@
+        pin_auth: Vec<u8>,
+        new_pin_enc: Vec<u8>,
+    ) -> Result<(), Ctap2StatusCode> {
+        if persistent_store.pin_hash().is_some() {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_INVALID);
+        }
+        let pin_decryption_key =
+            self.create_decryption_key_verified(key_agreement, &pin_auth, &new_pin_enc)?;
+        if !check_and_store_new_pin(persistent_store, &pin_decryption_key, new_pin_enc) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION);
+        }
+        persistent_store.reset_pin_retries();
+        Ok(())
+    }
+
+    fn process_change_pin(
+        &mut self,
+        rng: &mut impl Rng256,
+        persistent_store: &mut PersistentStore,
+        key_agreement: CoseKey,
+        pin_auth: Vec<u8>,
+        new_pin_enc: Vec<u8>,
+        pin_hash_enc: Vec<u8>,
+    ) -> Result<(), Ctap2StatusCode> {
+        if persistent_store.pin_retries() == 0 {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_BLOCKED);
+        }
+        let mut auth_param_data = new_pin_enc.clone();
+        auth_param_data.extend(&pin_hash_enc);
+        let pin_decryption_key =
+            self.create_decryption_key_verified(key_agreement, &pin_auth, &auth_param_data)?;
+        self.verify_pin_hash_enc(rng, persistent_store, &pin_decryption_key, pin_hash_enc)?;
+
+        if !check_and_store_new_pin(persistent_store, &pin_decryption_key, new_pin_enc) {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION);
+        }
+        self.pin_uv_auth_token = rng.gen_uniform_u8x32();
+        Ok(())
+    }
+
+    fn process_get_pin_token(
+        &mut self,
+        rng: &mut impl Rng256,
+        persistent_store: &mut PersistentStore,
+        key_agreement: CoseKey,
+        pin_hash_enc: Vec<u8>,
+    ) -> Result<AuthenticatorClientPinResponse, Ctap2StatusCode> {
+        if persistent_store.pin_retries() == 0 {
+            return Err(Ctap2StatusCode::CTAP2_ERR_PIN_BLOCKED);
+        }
+        let pk: crypto::ecdh::PubKey = CoseKey::try_into(key_agreement)?;
+        let shared_secret = self.key_agreement_key.exchange_x_sha256(&pk);
+
+        let token_encryption_key = crypto::aes256::EncryptionKey::new(&shared_secret);
+        let pin_decryption_key = crypto::aes256::DecryptionKey::new(&token_encryption_key);
+        self.verify_pin_hash_enc(rng, persistent_store, &pin_decryption_key, pin_hash_enc)?;
+
+        // Assuming PIN_TOKEN_LENGTH % block_size == 0 here.
+        let iv = [0u8; 16];
+        let mut blocks = [[0u8; 16]; PIN_TOKEN_LENGTH / 16];
+        for (i, item) in blocks.iter_mut().take(PIN_TOKEN_LENGTH / 16).enumerate() {
+            item.copy_from_slice(&self.pin_uv_auth_token[i * 16..(i + 1) * 16]);
+        }
+        cbc_encrypt(&token_encryption_key, iv, &mut blocks);
+        let mut pin_token = vec![];
+        for item in blocks.iter().take(PIN_TOKEN_LENGTH / 16) {
+            pin_token.extend(item);
+        }
```

Comment:
I think the same `iter.flatten().collect()` pattern applies here.

Reply:
Done